### ISSUE TITLE
source-postgres: connector fails if watermarks table doesn't exist

### ISSUE THREAD

**Issue by williamhbaker** [estuary/connectors#488]
With `source-postgres:v3`, the connector will fail if the specified watermarks table doesn't exist:

```
runTransactions: readMessage: connector failed (exit status: 2) with stderr:
{"address":"localhost:5432","database":"postgres","level":"info","msg":"initializing connector","slot":"flow_slot","time":"2023-01-04T18:56:50Z","user":"postgres"}
{"level":"info","msg":"discovering tables","time":"2023-01-04T18:56:51Z"}
{"level":"info","msg":"starting replication","publication":"flow_publication","slot":"flow_slot","startLSN":24272888,"time":"2023-01-04T18:56:53Z"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xa31a68]

goroutine 1 [running]:
github.com/estuary/connectors/sqlcapture.(*Capture).Run(0xc0000ce880, {0xeab968, 0xc000368c40})
	/builder/sqlcapture/capture.go:149 +0x788
github.com/estuary/connectors/sqlcapture.(*Driver).Pull(0x134f9c0, {0xeae450, 0xc00007e600})
	/builder/sqlcapture/main.go:180 +0x73f
github.com/estuary/connectors/source-boilerplate.pullCmd.Execute({{{0xeab968, 0xc000368c40}, {0xeacbe8, 0x134f9c0}, 0xc00030e660, {0x406c000f18, 0xc0000380a0}, 0xc000303500}}, {0xc00007a020, 0x0, ...})
	/builder/source-boilerplate/boilerplate.go:248 +0x111
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc000304fc0, {0xc00009c160, 0x2, 0x2})
	/go/pkg/mod/github.com/jessevdk/go-flags@v1.5.0/parser.go:335 +0x92e
github.com/jessevdk/go-flags.(*Parser).Parse(...)
	/go/pkg/mod/github.com/jessevdk/go-flags@v1.5.0/parser.go:191
github.com/estuary/connectors/source-boilerplate.RunMain({0xeacbe8, 0x134f9c0})
	/builder/source-boilerplate/boilerplate.go:53 +0x798
main.main()
	/builder/source-postgres/main.go:45 +0x27
```

This happens [here](https://github.com/estuary/connectors/blob/2cdfa473e2384373520a8db6ef6120a77c6ea6a7/sqlcapture/capture.go#L149), because `c.discovery[watermarks]` is `nil`. The watermarks table is not created until `WriteWatermark` is called (see [here](https://github.com/estuary/connectors/blob/2cdfa473e2384373520a8db6ef6120a77c6ea6a7/source-postgres/backfill.go#L100)). The previous `source-postgres:v2` connector did things a little differently and was able to create the watermarks table if it doesn't exist without a panic.

`source-mysql` and `source-postgres` share a lot of code via `sqlcapture`. One thing to note is that `source-mysql` requires a watermark table already exist and does not automatically create it if it doesn't (as far as I can tell, at least). So a possible resolution to this would be to make it a requirement of `source-postgres` that the watermarks table already exist. This seems non-ideal from a user perspective, but would be consistent with `source-mysql`. Otherwise we will need to reconfigure things to ensure that the watermarks table exists before it is activated.

**Comment by willdonnelly** [comment]
Several thoughts:
1. As best I can recall, there's no principled reason that `source-postgres` creates the watermarks table and `source-mysql` doesn't. We should make that behavior more consistent between the two, in either direction.
2. At the offending line of code, it's not actually a problem that `c.discovery[watermarks] == nil` (the replication stream logic handles "activating a table which doesn't currently exist" correctly), the problem is solely because we try to do `c.discovery[watermarks].PrimaryKey` to get the name of the key column. That information could be sourced in an entirely different way, whether that's by adding another helper method to the database driver interface or even by just making sure the key column always has the same name on the various SQL databases we support and hard-coding it.
3. A really easy solution is probably to just an an explicit check at this spot, verifying that the watermarks table existed at the the point in time where discovery was performed. If we still want it to be auto-created at startup, that should probably be moved into the database connection+setup code rather than being done automagically in `WriteWatermark()` anyway

**Comment by willdonnelly** [comment]
After thinking a bit longer I'm starting to remember why there's a discrepancy between Postgres and MySQL here. The issue is that Postgres has a default schema `public` so it's reasonable to just default to `public.flow_watermarks` if the user hasn't specified something different. But MySQL has no default schema, and while our recommended setup is:

```
CREATE DATABASE IF NOT EXISTS flow;
CREATE TABLE IF NOT EXISTS flow.watermarks (slot INTEGER PRIMARY KEY, watermark TEXT);
```

automagically creating a database felt a bit dicier than merely autocreating a single table, so I didn't do it when I first sat down to write `source-mysql`.

However I think it would actually be reasonable to start doing that, because the name `flow.watermarks` in the configuration can be seen as the user giving us the implicit go-ahead for that `CREATE DATABASE IF NOT EXISTS`.

**Comment by willdonnelly** [comment]
TL;DR recommended solution:
1. In `source-postgres`, move the `CREATE TABLE IF NOT EXISTS` thing out of the `WriteWatermark` method and put it in the database-specific connection logic. Possibly also move the `CREATE REPLICATION SLOT` stuff to there as well, so all of this automagic creating-stuff helper behavior happens in one spot.
2. In `source-mysql`, add an equivalent `CREATE DATABASE IF NOT EXISTS; CREATE TABLE IF NOT EXISTS` sequence right after database connection, so that Postgres and MySQL behavior are as similar as possible.
3. In `sqlcapture`, add a sanity check right before the whole `ActivateStream(watermarks)` thing, so that just in case something goes wrong we produce a useful error message rather than a stack trace.

I can take care of this sometime in the next week or two if it's not urgent.

**Comment by williamhbaker** [comment]
Tested this out today and it is fixed - doing a capture on a postgres db without a pre-created watermarks table now works as expected, so I'll close this out.

**Comment by willdonnelly** [comment]
I've filed https://github.com/estuary/connectors/issues/537 for the last remaining bit of work discussed here: making MySQL autocreate the watermarks table like Postgres and SQL Server captures do.
